### PR TITLE
Adding unit tests

### PR DESCRIPTION
**Pull request by ktonon**
```shell
> npm install -g elm-test
> elm test
Success! Compiled 0 modules.
Successfully generated /dev/null
Success! Compiled 1 module.
Successfully generated /Users/kevin/bc/elm-memo/elm-stuff/generated-code/elm-community/elm-test/elmTestOutput.js

elm-test
--------

Running 8 tests. To reproduce these results, run: elm-test --fuzz 100 --seed 1464326693


TEST RUN PASSED

Duration: 8 ms
Passed:   8
Failed:   0
```

### CODE REVIEW

File: tests/MemoTests.elm [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+                    Expect.equal
+                        Nothing
+                        (squareMemo 0)
+            ]
+        , describe "randomInt"
+            [ test "returns different values each time" <|
+                \_ ->
+                    Expect.notEqual
+                        (randomInt upper)
+                        (randomInt upper)
```
Comment (ktonon) [inline tests/MemoTests.elm:39]: native non-pure function

File: tests/MemoTests.elm [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            [ test "returns a value if argument in range" <|
+                \_ ->
+                    Expect.notEqual
+                        (randomIntMemo upper |> Maybe.withDefault -1)
+                        -1
+            , test "returns the same value each time" <|
+                \_ ->
+                    Expect.equal
+                        (randomIntMemo upper)
+                        (randomIntMemo upper)
```
Comment (ktonon) [inline tests/MemoTests.elm:51]: can test that source function is not called twice

**Comment by jvoigtlaender** [comment]
Pretty cool, thanks for doing this!

I have to ask: are you aware that this library will most likely be dead with the next release of Elm, since the native elm-lang/lazy library it depends on is said to be axed?

**Comment by jvoigtlaender** [comment]
Do you have a practical use case for elm-memo? Maybe that would convince Evan to preserve elm-lang/lazy.

**Comment by ktonon** [comment]
No, wasn't aware of the plan to axe elm-lang/lazy. Is it just elm-lang/lazy, or native modules altogether? Memoization would be impossible without native modules right?

My use case is a server side HTTP processing pipeline framework. The pipelines process connections (`Conn`), which are composed of:

* request
* response
* app specific model
* app specific config

Pipelines are made up of `Plug` functions. I'm thinking of using memoization for the branching functions called routers, which currently have the signature `Conn -> Plug` ([for example](https://github.com/ktonon/elm-serverless/blob/master/demo/src/API.elm#L46)). These functions are potentially expensive.

Not sure yet if its a good idea. As they are now, the argument space of these functions is too big to benefit from memoization, so I'd have to redefine them with a more restricted argument space.

In any case, nice library! Hope it survives the next elm release.

**Comment by jvoigtlaender** [comment]
> No, wasn't aware of the plan to axe elm-lang/lazy. Is it just elm-lang/lazy, or native modules altogether? Memoization would be impossible without native modules right?

About the plan to axe elm-lang/lazy, this is where it is mentioned: https://github.com/elm-lang/lazy/commit/c9c3f8525d22978cd7ee6e463ebf208f30fa1f91

You are right that implementing memoization depends on native modules. As such, since a normal library author cannot upload native modules to the package repository, it will be impossible to preserve `elm-memo` once `elm-lang/lazy` is not provided anmyore. 

I don't think that native modules will be axed altogether. But like until now, it will be very limited who can upload native modules as part of a library. Essentially it will still be the case that only Evan can do that. So with his decision to no longer support `elm-lang/lazy`, that door is closed.

Concerning use cases for memoization, Evan writes in the commit referenced above: "In all the cases in Elm that I have heard of that use this library [`elm-lang/lazy`], folks only really needed *delayed computation* and ended up with simpler code when they went that way." So essentially he is saying that he hasn't seen uses of memoization (sharing rather than just delaying computation) that he considers worthwhile. Of course, `elm-memo` needs more than just delaying computation, but I didn't create `elm-memo` for a specific practical use case; I created it during a discussion on the mailing list, to explain memoization and show how it can be done in "Elm plus `elm-lang/lazy`". A practical use case involving `elm-memo` would be new information to Evan (that might change his stance as expressed in the above quote).

**Comment by ktonon** [comment]
Thanks for the detailed summary. It's too bad about `elm-lang/lazy`. I would prefer a non-pure implementation of a function memoizer, but I also don't want to be going against the current. I've created a pure implementation at [ktonon/elm-memo-pure](http://package.elm-lang.org/packages/ktonon/elm-memo-pure/latest/Memo). It works by giving you back a new `Memo` whenever you call the memoized function.

**Pull request merged**